projects/client/RabbitMQ.Client/src/client/impl/Connection.cs

```diff
@@ -1356,4 +1357,1303 @@ private static uint NegotiatedMaxValue(uint clientValue, uint serverValue)
+        }
+
+        public void MainLoop()
+        {
+            try
+            {
+                bool shutdownCleanly = false;
+                try
+                {
+                    while (m_running)
+                    {
+                        try
+                        {
+                            MainLoopIteration();
+                        }
+                        catch (SoftProtocolException spe)
+                        {
+                            QuiesceChannel(spe);
+                        }
+                    }
+                    shutdownCleanly = true;
+                }
+                catch (EndOfStreamException eose)
+                {
+                    // Possible heartbeat exception
+                    HandleMainLoopException(new ShutdownEventArgs(
+                        ShutdownInitiator.Library,
+                        0,
+                        "End of stream",
+                        eose));
+                }
+                catch (HardProtocolException hpe)
+                {
+                    shutdownCleanly = HardProtocolExceptionHandler(hpe);
+                }
+#if !NETFX_CORE
+                catch (Exception ex)
+                {
+                    HandleMainLoopException(new ShutdownEventArgs(ShutdownInitiator.Library,
+                        Constants.InternalError,
+                        "Unexpected Exception",
+                        ex));
+                }
+#endif
+
+                // If allowed for clean shutdown, run main loop until the
+                // connection closes.
+                if (shutdownCleanly)
+                {
+#pragma warning disable 0168
+                    try
+                    {
+                        ClosingLoop();
+                    }
+#if NETFX_CORE
+                    catch (Exception ex)
+                    {
+                        if (SocketError.GetStatus(ex.HResult) != SocketErrorStatus.Unknown)
+                        {
+                            // means that socket was closed when frame handler
+                            // attempted to use it. Since we are shutting down,
+                            // ignore it.
+                        }
+                        else
+                        {
+                            throw ex;
+                        }
+                    }
+#else
+                    catch (SocketException se)
+                    {
+                        // means that socket was closed when frame handler
+                        // attempted to use it. Since we are shutting down,
+                        // ignore it.
+                    }
+#endif
+#pragma warning restore 0168
+                }
+
+                FinishClose();
+            }
+            finally
+            {
+                m_appContinuation.Set();
+            }
+        }
+
+        public void MainLoopIteration()
+        {
+            Frame frame = m_frameHandler.ReadFrame();
```

Comment:
This part would benefit greatly from async / await as the thread is waiting to receive from the network serially